ext/fg/js/frame-offset-forwarder.js

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2020  Alex Yatskov <alex@foosoft.net>
+ * Author: Alex Yatskov <alex@foosoft.net>
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+/* global
+ * apiForward
+ */
+
+class FrameOffsetForwarder {
+    constructor() {
+        this._forwardFrameOffset = window !== window.parent ?
+            this._forwardFrameOffsetParent.bind(this) :
+            this._forwardFrameOffsetOrigin.bind(this);
+
+        this._windowMessageHandlers = new Map([
+            ['getFrameOffset', ({offset, uniqueId}, e) => { return this._onGetFrameOffset(offset, uniqueId, e); }]
+        ]);
+
+        window.addEventListener('message', this.onMessage.bind(this), false);
+    }
+
+    async applyOffset(x, y) {
+        const uniqueId = yomichan.generateId(16);
+
+        let frameOffsetResolve = null;
+        const frameOffsetPromise = new Promise((resolve) => (frameOffsetResolve = resolve));
+
+        const runtimeMessageCallback = ({action, params}, sender, callback) => {
+            if (action === 'frameOffset' && isObject(params) && params.uniqueId === uniqueId) {
+                chrome.runtime.onMessage.removeListener(runtimeMessageCallback);
+                callback();
+                frameOffsetResolve(params);
+                return false;
+            }
+        };
+        chrome.runtime.onMessage.addListener(runtimeMessageCallback);
+
+        window.parent.postMessage({
+            action: 'getFrameOffset',
+            params: {
+                uniqueId,
+                offset: [x, y]
+            }
+        }, '*');
+
+        const {offset} = await frameOffsetPromise;
+        return offset;
+    }
+
+    onMessage(e) {
+        const {action, params} = e.data;
+        const handler = this._windowMessageHandlers.get(action);
+        if (typeof handler !== 'function') { return; }
+        handler(params, e);
+    }
+
+    _onGetFrameOffset(offset, uniqueId, e) {
+        let sourceFrame = null;
+        for (const frame of document.querySelectorAll('frame, iframe:not(.yomichan-float)')) {
+            if (frame.contentWindow !== e.source) { continue; }
+            sourceFrame = frame;
+            break;
+        }
+        if (sourceFrame === null) { return; }
```

Comment:
May want this to always send a response, even if the frame couldn't be found.

Example case: if the iframe is in a shadow DOM, `document.querySelector` won't return anything, and thus `runtimeMessageCallback` will never be invoked. Repeated calls to `applyOffset` will then result in a memory leak since event callbacks are being added without ever being removed.

Reply:
I might add a timeout to the promise in `applyOffset`. I'm also considering creating an abstraction for the temporary listener as you mentioned, and it could take timeout as an argument.

Reply:
https://github.com/FooSoft/yomichan/commit/9fdcedb4001a5e86cc7e648c61bd9961a202fa54 adds a timeout that rejects everything that takes more than 30 seconds. The timeouts can be adjusted based on how critical the result is.

Reply:
This should probably still return some sort of "unknown" indicator and the timeout should be treated as a worst-case scenario. The `getFrameOffset` invocation is likely to happen a lot due to mouse movement, and if the timeout is 5 seconds, a very large message queue can build up.

If an "unknown" indicator is returned, this would also enable the child window to decide to skip the call in the future, since if it's unknown once, it's likely to be unknown forever. Not necessary to implement this now, but I think it's good to have that as possibility.

I think we should also start putting together a HTML page in the test folder with some of the odd edge cases we encounter / can think of, such as the iframe in a shadow DOM and the fullscreen issue.

Reply:
https://github.com/FooSoft/yomichan/commit/204ae40869f904a34d272204b9ba17d345686752 also relevant here

Reply:
> This should probably still return some sort of "unknown" indicator and the timeout should be treated as a worst-case scenario. The getFrameOffset invocation is likely to happen a lot due to mouse movement, and if the timeout is 5 seconds, a very large message queue can build up.
> 
> If an "unknown" indicator is returned, this would also enable the child window to decide to skip the call in the future, since if it's unknown once, it's likely to be unknown forever. Not necessary to implement this now, but I think it's good to have that as possibility.

Addressed in https://github.com/FooSoft/yomichan/pull/417/commits/204ae40869f904a34d272204b9ba17d345686752. Updating the timestamp should also partially help in cases when it's always unknown by throttling the checks to one per second.

> I think we should also start putting together a HTML page in the test folder with some of the odd edge cases we encounter / can think of, such as the iframe in a shadow DOM and the fullscreen issue.

I think this is a good idea, I'm currently developing by repeatedly opening a Chrome profile with some of these test cases found in the wild open. There's some delay opening them and I got rate limited more than once :joy: